### PR TITLE
🔨 Handle autocomplete project suggestions effect

### LINKED ISSUE #257: Make the UpdateAutocompleteSuggestionsEffect return project/task suggestions
## Description
We need to update the effect that calculates autocomplete suggestions to make it handle project and task suggestions.
What we need to do is mimic the way the Xamarin app does the search. We find if the caret is currently placed in a word that starts with an @. If it is, we use that word to suggest projects and tasks, matching against the project/task name. Check the Xamarin implementation for more details
The list of suggestions returned must always include a `CreateProject` autocomplete suggestion, with the word that's included in the @ as the suggested name.

## Definition of done
- [ ] The effect is updated to return a list of project/task/create project autocomplete suggestions
- [ ] Tests ensure that the behavior is correct

### PR DESCRIPTION
**Pull request by daividssilverio**
### Summary
This adds handling for the auto complete suggestions for project queries (the ones that start with '@').

### Technical considerations
The `lastSubstringFromAnyTokenToPosition` extension currently works, as the name suggests, considering the last token added to the query string. Is that what we want?

So if the user types "@ something", projects would be searched, but if they type "@ something #", tags would suggestions would start appearing instead.

### Why is it done like this and not some other way?
Feel free to make suggestions. The token handling in special could receive some opinions on how it should work.

### Relationships
Closes #257 

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [ ] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

I'm lacking ideas of tests to add. I'm welcoming ideas for tests.

### Pointers
Review by commit;
Please check the tests.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

## Merge permissions
🦑 it

### CODE REVIEW

File: timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt [hunk 1]
```diff
@@ -53,13 +63,42 @@ class UpdateAutocompleteSuggestionsEffect(
...
+    }
+
+    private fun fetchTasksSuggestionsFor(query: String): List<AutocompleteSuggestion> {
+        val words = query.split(wordSeparator)
+        return words.fold(tasks.values) { tasks, word ->
+            tasks.filter { task ->
+                task.name.contains(word, true)
+            }
+        }.map(AutocompleteSuggestion::Task)
+    }
```
Comment (semanticer) [inline timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt:103]: These fetch methods are starting to look very similar. What about defining helper method to hide some inner plumbing in order to simplify our fetch methods and make them more readable. 

```suggestion
    private fun fetchTasksSuggestionsFor(query: String): List<AutocompleteSuggestion> =
        fetchSuggestionsFor(query, tasks.values, AutocompleteSuggestion::Task) { task, word ->
            task.name.contains(word, true)
        }
        
    private fun <T> fetchSuggestionsFor(
        query: String,
        possibleSuggestions: Collection<T>,
        toSuggestion: (T) -> AutocompleteSuggestion,
        predicate: (T, String) -> Boolean
    ): List<AutocompleteSuggestion> {
        val words = query.split(wordSeparator)
        return words.fold(possibleSuggestions) { remainingPossibleSuggestions, word ->
            remainingPossibleSuggestions.filter { predicate(it, word) }
        }.map(toSuggestion)
    }
```

File: timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt [hunk 2]
```diff
@@ -22,15 +23,24 @@ class UpdateAutocompleteSuggestionsEffect(
...
-            val suggestions = fetchTimeEntrySuggestionsFor(words)
+            if (query.isBlank())
+                return@withContext StartEditAction.AutocompleteSuggestionsUpdated(emptyList())
+
+            val actualQuery = query.lastSubstringFromAnyTokenToPosition(shortcutTokens, cursorPosition)
+            val suggestions = when {
+                actualQuery.startsWith(projectShortCut) ->
+                    fetchProjectSuggestionsFor(actualQuery.drop(1)) + fetchTasksSuggestionsFor(actualQuery.drop(1))
+                else -> fetchTimeEntrySuggestionsFor(actualQuery)
+            }
```
[the remark this answers is not in the document]
Comment (semanticer) [reply timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt:39]: Maybe we should create an even more universal version of `lastSubstringFromAnyTokenToPosition` which returns all `token -> text` combinations in order of appliance and just pick the last one in this case. 
But for example in https://github.com/toggl/android/issues/254 we have this requirement _"Selecting a project suggestion should also remove the word that starts with an @ from the description (do so by checking the cursor position as well)"_ so we might want to use something like `lastSubstringFromAnyTokenToPosition` but pick based on the token, not based on the position? I'm not sure yet.

File: timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt [hunk 3]
```diff
@@ -53,13 +63,42 @@ class UpdateAutocompleteSuggestionsEffect(
...
+
         return words.fold(timeEntries.values) { timeEntries, word ->
             timeEntries.filter { timeEntry ->
                 timeEntry.description.contains(word, true) ||
-                timeEntry.projectOrClientNameContains(word) ||
-                timeEntry.tagNamesContain(word) ||
-                timeEntry.taskNameContains(word)
+                    timeEntry.projectOrClientNameContains(word) ||
+                    timeEntry.tagNamesContain(word) ||
+                    timeEntry.taskNameContains(word)
```
[the remark this answers is not in the document]
Comment (daividssilverio) [reply timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt:73]: I just executed auto format.

File: timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt [hunk 4]
```diff
@@ -53,13 +63,42 @@ class UpdateAutocompleteSuggestionsEffect(
...
+        val words = query.split(wordSeparator)
+        return sequence {
+            yield(AutocompleteSuggestion.CreateProject(query))
+            words.fold(projects.values) { projects, word ->
+                projects.filter { project ->
+                    project.name.contains(word, true) ||
+                        project.clientNameContains(word)
+                }
+            }.map(AutocompleteSuggestion::Project).let { yieldAll(it) }
+        }.toList()
```
[the remark this answers is not in the document]
Comment (daividssilverio) [reply timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt:93]: Not really.

File: timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (daividssilverio) [reply timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt:39]: 🤔 Are you thinking about something like this?

```kotlin
val tokenQueries = "my cool @amazing description @with cool #tags and @stuff".lastSubstringFromAnyTokenToPosition('@', '#')
tokenQueries shouldBe listOf(
  '@' to "amazing description @with cool #tags and @stuff",
  '@' to "with cool #tags and @stuff",
  '#' to "tags and @stuff",
  '@' to "stuff"
)
```

In this example, we would only match suggestions for the last matched token/query. We would search for projects with the query `stuff`.

When it's time to replace the string, we would replace only the last token and the query after it up to the cursor position. The result would be something like `my cool @amazing description @with cool #tags and `

If we consider the cursor position as well, in the same example:
```
val tokenQueries = "my cool @amazing description <<<cursor here>>>@with cool #tags and @stuff".lastSubstringFromAnyTokenToPosition('@', '#', cursorPosition)
tokenQueries shouldBe listOf(
  '@' to "amazing description "
)
```

If the suggestion is tapped, the resulting description would look like:
`my cool @with cool #tags and @stuff"`

**Pull request merged**